# prepare_all_gold_standards.r
# This script prepares the gold standard sensitive genes
# Gold standard sensitive genes are identified as the genes that a cell is particularly sensitive to losing relative to other cells of the same lineage
# If global alpha is supplied, also considers highly sensitive genes even if that cell is not an outlier


# Load Required Packages
message("Loading Packages")
suppressPackageStartupMessages (library(optparse, quietly = T))
suppressPackageStartupMessages (library(tidyverse, quietly = T))
suppressPackageStartupMessages (library(readxl, quietly = T))
suppressPackageStartupMessages (library(data.table, quietly = T))

# Handling input arguments
option_list = list(
  make_option(c("-l", "--local_alpha"), type="double", default=0.1, 
              help="Local (within a cell type) alpha significance threshold to be used (default = 0.1)", metavar ="Local Alpha"),
  make_option(c("-g", "--global_alpha"), type="double", default=0.1, 
              help="Global (including all cell types) alpha significance threshold to be used (default = 0.1)", metavar ="Global Alpha"),
  make_option(c("-m", "--method"), type="character", default="picklesv3", 
              help="Method of identifying gold standards", metavar ="Method")
); 

opt_parser = OptionParser(option_list=option_list);
opt = parse_args(opt_parser);

local_alpha <- opt$local_alpha
global_alpha <- opt$global_alpha
method <- opt$method

#############################
# Functions
###############################

# Initially found outliers based on z-score thresholds
# Problematically, this approach meant that the "most outlying" cells were always detected as outliers
# Changed to GESD (below) to allow for statistical testing for outliers

get_z_scores <- function(Matrix){
  Sd   <- apply(Matrix, 2, function(x) sd(x, na.rm=T))
  Mean <- apply(Matrix, 2, function(x) mean(x, na.rm=T))
  centered <- suppressWarnings(t(t(Matrix) - Mean))
  z_scores <- suppressWarnings(t(t(centered) / Sd))
  return(z_scores)
}

# Applying Generalized Extreme Studentized Deviate (GEDS) Test
# R implementation taken from https://github.com/raunakms/GESD
# runGESD.R
source("scripts/runGESD.R")

get_outliers <- function(Matrix, alpha){
  result <- apply(Matrix, 2, function(x) gesd(x, alpha = alpha, value.zscore = "NO", r = nrow(Matrix)/4))
}

# Note, the above test assumes normally distributed data, whihc is likely not true in this case.
# Could consider changing to using Median Absolute Deviation

#############################
# Get cell list
###############################
# This cell list is generated by the choose_cells.r script

cell_list <- read_csv("data/cell_list.csv")

#############################
# Get cell line information
#############################

CCLE_sample_info <- read_csv("data/CCLE/Model.csv") %>%
  dplyr::select(DepMap_ID=ModelID, lineage=OncotreeLineage, cell_ID=StrippedCellLineName) %>%
  mutate(lineage=gsub(" |/","_",lineage)) %>%
  filter(cell_ID %in% cell_list$cell_ID) %>%
  unique()

#############################
# Get Expressed Genes
#############################

CCLE_tpm <- data.table::fread("data/CCLE/OmicsExpressionProteinCodingGenesTPMLogp1.csv") %>%
  # Replacing DepMap IDs with cell names
  dplyr::rename(DepMap_ID = 1) %>%
  dplyr::right_join(dplyr::select(CCLE_sample_info, DepMap_ID, cell_ID), by = "DepMap_ID") %>%
  na.omit() %>%
  dplyr::select(-DepMap_ID) %>% dplyr::relocate(cell_ID) %>%
  data.table::transpose(make.names = "cell_ID", keep.names = "gene_ID") %>%
  # Fixing gene_IDs by removing everything after a space
  mutate(gene_ID = gsub(" .*","",gene_ID)) %>%
  arrange(gene_ID) %>%
  column_to_rownames("gene_ID") %>%
  # Trimming genes with low reads
  
  # Filtering low expression based on TPM as this accounts for gene length
  # Cells contain approx 200 000 mRNA at any time, therefore tpm of 5 ~ 1 transcript per cell
  # Therefore, tpm of 5 is beging used a cutoff to select genes that are being expressed
  # Data is in log2(tpm+1), therefore filtering for expression abofe log2(5+1)
  dplyr::filter(rowMeans(dplyr::select(., where(is.numeric))) > log2(5+1)) %>%
  as.data.frame()

genes <- rownames(CCLE_tpm)
rm(CCLE_tpm)




if(method == "gesd"){

#############################
# Read In Data
#############################
message("Reading in data")


gene_effect <- fread("data/CCLE/CRISPRGeneEffect.csv") %>%
  as.data.frame() %>%
  dplyr::rename(DepMap_ID=V1) %>%
  inner_join(CCLE_sample_info[c("DepMap_ID","cell_ID")], by = "DepMap_ID") %>%
  dplyr::select(-DepMap_ID) %>%
  column_to_rownames("cell_ID")

colnames(gene_effect) <- gsub(" \\(.*","", colnames(gene_effect))
gene_effect <- gene_effect[!duplicated(colnames(gene_effect))]
gene_effect <- gene_effect[colnames(gene_effect) %in% genes]


gene_effect_global_outliers <- gene_effect %>%
  # First adding a row of gene_effect means for each gene across the cells
  rbind(
    colMeans(gene_effect, na.rm = T) %>% 
      matrix(nrow = 1, dimnames = list("global_gene_effect_mean",colnames(gene_effect)))
  ) %>%
  # Transposing data
  rownames_to_column("cell_ID") %>%
  data.table::transpose(make.names = "cell_ID", keep.names = "gene_ID") %>%
  pivot_longer(cols = -c(gene_ID, global_gene_effect_mean), names_to = "cell_ID", values_to = "gene_effect") %>%
  # Then joining with z-scores and calculated outliers
  inner_join(
    get_outliers(gene_effect,global_alpha) %>%
    as.data.frame() %>%
    slice(-1) %>%
    rownames_to_column("cell_ID") %>%
    pivot_longer(cols = -cell_ID, names_to = "gene_ID", values_to = "global_outlier_rank") %>%
    mutate(global_outlier = global_outlier_rank > 0),
    by = c("gene_ID", "cell_ID")
  ) %>%
  # Defining dependent genes as outliers with values < mean
  mutate(global_dependent = ifelse(global_outlier & gene_effect < global_gene_effect_mean, TRUE, FALSE))



gene_effect_global_z_scores <- get_z_scores(gene_effect) %>%
  as.data.frame() %>%
  rownames_to_column("cell_ID") %>%
  pivot_longer(cols = -cell_ID, names_to = "gene_ID", values_to = "global_z_score")



# Getting GDSC versions of cell_IDs
GDSC_IDs <- read_xlsx("data/GDSC/Cell_Lines_Details.xlsx",sheet = "Cell line details") %>%
  dplyr::select(GDSC_ID = `Sample Name`, COSMIC_ID = `COSMIC identifier`) %>%
  mutate(cell_ID = toupper(gsub("[[:punct:], ]", "", GDSC_ID)))

# Getting all possible synonyms of each drug and choosing a primary name to map all synonyms back to
drug_synonyms <- read.csv("data/GDSC/screened_compounds_rel_8.4.csv") %>%
  dplyr::select(DRUG_NAME, SYNONYMS) %>%
  mutate(synonym_0 = DRUG_NAME,
         n_synonyms = nchar(gsub("[^,]","", SYNONYMS))+1)
drug_synonyms <- drug_synonyms %>%
  separate(SYNONYMS, sep = ",", into = paste("synonym", 1:max(drug_synonyms$n_synonyms, na.rm = T))) %>%
  dplyr::select(-n_synonyms) %>%
  pivot_longer(cols = -DRUG_NAME, names_to = "category", values_to = "synonyms") %>%
  dplyr::select(primary_name = DRUG_NAME, synonyms) %>%
  na.omit() %>%
  mutate(primary_name = toupper(gsub("[[:punct:], ]", "", primary_name)),
         synonyms = toupper(gsub("[[:punct:], ]", "", synonyms))) %>%
  unique() %>%
  filter(synonyms != "")

# Reading in drug sensitivity data (firstly, GDSC1)
GDSC1 <- read_xlsx("data/GDSC/GDSC1_fitted_dose_response_24Jul22.xlsx") %>%
  dplyr::select(DATASET,COSMIC_ID, DRUG_ID, DRUG_NAME, LN_IC50) %>%
  inner_join(GDSC_IDs, by = "COSMIC_ID") %>% dplyr::select(-GDSC_ID, -COSMIC_ID) %>%
  dplyr::relocate(cell_ID) %>%
  unique() %>%
  na.omit()

GDSC2 <- read_xlsx("data/GDSC/GDSC2_fitted_dose_response_24Jul22.xlsx") %>%
  dplyr::select(DATASET,COSMIC_ID, DRUG_ID, DRUG_NAME, LN_IC50) %>%
  inner_join(GDSC_IDs, by = "COSMIC_ID") %>% dplyr::select(-GDSC_ID, -COSMIC_ID) %>%
  dplyr::relocate(cell_ID) %>%
  unique() %>%
  na.omit()

GDSC <- rbind(GDSC1, GDSC2) %>%
  mutate(DRUG_NAME = toupper(gsub("[[:punct:], ]", "", DRUG_NAME)))

rm(GDSC1, GDSC2)

# Get drug-gene ineteractions from DGIdb

DGIdb <- fread("data/DGIdb/interactions.tsv", sep = "\t")

drug_types <- DGIdb$interaction_types %>% unique() %>% sort() 
# Selecting only inhibitory interaction types
drug_types <- drug_types[c(7,10,11,12,13,14,16,18,19,20,22,25)]

write_csv(data.frame(drug_types),"data/DGIdb/selected_drug_types.csv")

DGIdb <- DGIdb %>% 
  pivot_longer(cols = drug_claim_name:drug_name, names_to = "name_category", values_to = "drug_name") %>%
  mutate(drug_name = toupper(gsub("[[:punct:], ]", "", drug_name))) %>%
  inner_join(drug_synonyms, by = c("drug_name"="synonyms"), multiple="all") %>%
  dplyr::select(gene_name, drug=primary_name,interaction_types) %>%
  filter(interaction_types %in% drug_types) %>%
  unique() %>%
  na.omit()

GDSC <- GDSC %>%
  left_join(DGIdb, by = c("DRUG_NAME"="drug"), multiple="all") %>%
  filter(gene_name %in% genes)

# At this point, we can quickly take the full list of possible genes
total_genes <- union(GDSC$gene_name, colnames(gene_effect))

GDSC <- GDSC %>%
  dplyr::select(DATASET, cell_ID, DRUG_NAME, gene_name, LN_IC50) %>%
  unique() %>%
  mutate(target_ID = paste(DATASET,DRUG_NAME, gene_name, sep = "_")) %>%
  dplyr::select(cell_ID, target_ID, LN_IC50) %>%
  # Taking average of replicate data  
  group_by(cell_ID, target_ID) %>%
  summarise(LN_IC50 = mean(LN_IC50)) %>%
  pivot_wider(names_from = "target_ID", values_from = "LN_IC50") %>%
  column_to_rownames("cell_ID")

GDSC_global_outliers <- GDSC %>%
  # First adding a row of gene_effect means for each gene across the cells
  rbind(
    colMeans(GDSC, na.rm = T) %>% 
      matrix(nrow = 1, dimnames = list("global_LN_IC50_mean",colnames(GDSC)))
  ) %>%
  # Transposing data
  rownames_to_column("cell_ID") %>%
  data.table::transpose(make.names = "cell_ID", keep.names = "target_ID") %>%
  pivot_longer(cols = -c(target_ID, global_LN_IC50_mean), names_to = "cell_ID", values_to = "LN_IC50") %>%
  # Then joining with z-scores and calculated outliers
  inner_join(
    get_outliers(GDSC,global_alpha) %>%
      as.data.frame() %>%
      slice(-1) %>%
      rownames_to_column("cell_ID") %>%
      pivot_longer(cols = -cell_ID, names_to = "target_ID", values_to = "global_outlier_rank") %>%
      mutate(global_outlier = global_outlier_rank > 0),
    by = c("target_ID", "cell_ID")
  ) %>%
  separate(target_ID, into = c("dataset", "drug_name", "gene_ID")) %>%
  # Defining sensitive genes as outliers with values < mean
  mutate(global_sensitive = ifelse(global_outlier & LN_IC50 < global_LN_IC50_mean, TRUE, FALSE))


for(lin in sort(unique(CCLE_sample_info$lineage))){
  
  message(paste0("Identifying gold standards for ", lin, " cells"))
  
  lineage_cells <- CCLE_sample_info %>% filter(lineage==lin) %>% dplyr::select(-lineage)
  
  if(length(which(lineage_cells$cell_ID %in% rownames(gene_effect))) < 4 & length(which(lineage_cells$cell_ID %in% rownames(GDSC))) < 4){
    next
  }
  
  
  
  if(length(which(lineage_cells$cell_ID %in% rownames(gene_effect))) >= 4){
      
    
    
    #############################
    # Gene Dependency
    #############################
    
    # Probabilities of dependency are calculated for each gene score in a cell line as the probability 
    # that score arises from the distribution of essential gene scores rather than nonessential gene scores. 
    # “Probability of dependency” is essentially modeling the gene effect as either being sampled from the 
    # distribution arising from “essential” genes or “non-essential” genes, and being a probability, the values range from 0 to 1.
    
    lin_gene_effect_matrix <- gene_effect %>%
      rownames_to_column("cell_ID") %>%
      filter(cell_ID %in% lineage_cells$cell_ID) %>%
      column_to_rownames("cell_ID")
    
    
    lin_gene_effect <- lin_gene_effect_matrix %>%
      # First adding a row of gene_effect means for each gene across the cells
      rbind(
        colMeans(lin_gene_effect_matrix, na.rm = T) %>% 
          matrix(nrow = 1, dimnames = list("local_gene_effect_mean",colnames(lin_gene_effect_matrix)))
        ) %>%
      # Transposing data
      rownames_to_column("cell_ID") %>%
      data.table::transpose(make.names = "cell_ID", keep.names = "gene_ID") %>%
      pivot_longer(cols = -c(gene_ID, local_gene_effect_mean), names_to = "cell_ID", values_to = "gene_effect") %>%
      # Then joining with z-scores and calculated outliers
      inner_join(
        get_outliers(lin_gene_effect_matrix,local_alpha) %>%
          as.data.frame() %>%
          slice(-1) %>%
          rownames_to_column("cell_ID") %>%
          pivot_longer(cols = -cell_ID, names_to = "gene_ID", values_to = "local_outlier_rank") %>%
          mutate(local_outlier = local_outlier_rank > 0),
        by = c("gene_ID", "cell_ID")
      ) %>%
      # Defining dependent genes as outliers with values < mean
      mutate(local_dependent = ifelse(local_outlier & gene_effect < local_gene_effect_mean, TRUE, FALSE)) %>%
      mutate(dependent = local_dependent) %>%
      dplyr::select(cell_ID, gene_ID, gene_effect, local_gene_effect_mean, local_outlier_rank, local_outlier, local_dependent, dependent)
    
    if(global_alpha != FALSE){
      lin_gene_effect <- lin_gene_effect %>%
        left_join(gene_effect_global_outliers, by = c("cell_ID","gene_ID","gene_effect")) %>%
        mutate(dependent = ifelse(global_dependent, TRUE, dependent))
      
    }
    
    #rm(lin_gene_effect_matrix)
    
    indiv_gene_effect_local_z_scores <- get_z_scores(lin_gene_effect_matrix) %>%
      as.data.frame() %>%
      rownames_to_column("cell_ID") %>%
      pivot_longer(cols = -cell_ID, names_to = "gene_ID", values_to = "local_z_score") %>%
      mutate(lineage = lin)
  
  }
  
  if(length(which(lineage_cells$cell_ID %in% rownames(GDSC))) >= 4){
  
    #############################
    # Drug Sensitivity
    #############################
    
    # Available from GDSC with z-scores (one cell vs all other cells) as well as ln(IC50)
    # We use provided z-scores for static thresholds and calculate new lineage-specific z-scores to identify outliers
    
  
    
    lin_GDSC_matrix <- GDSC %>%
      rownames_to_column("cell_ID") %>%
      filter(cell_ID %in% lineage_cells$cell_ID) %>%
      column_to_rownames("cell_ID")
    
    lin_drug_sensitivity <- lin_GDSC_matrix %>%
      
      # First getting original sentitivty and gene-wise sentitivty means
      rbind(colMeans(lin_GDSC_matrix, na.rm = T) %>% 
              matrix(nrow = 1, dimnames = list("local_LN_IC50_mean",colnames(lin_GDSC_matrix)))
            ) %>%
      # Transposing data
      rownames_to_column("cell_ID") %>%
      data.table::transpose(make.names = "cell_ID", keep.names = "target_ID") %>%
      pivot_longer(cols = -c(target_ID, local_LN_IC50_mean), names_to = "cell_ID", values_to = "LN_IC50") %>%
      # Then joining with z-scores and calculated outliers
      inner_join(
        get_outliers(lin_GDSC_matrix,local_alpha) %>%
          as.data.frame() %>%
          slice(-1) %>%
          rownames_to_column("cell_ID") %>%
          pivot_longer(cols = -cell_ID, names_to = "target_ID", values_to = "local_outlier_rank") %>%
          mutate(local_outlier = local_outlier_rank > 0),
        by = c("target_ID", "cell_ID")
      ) %>%
      #mutate(IC50 = round(exp(IC50), 2), local_IC50_mean = round(exp(local_IC50_mean), 2)) %>%
      # Defining sensitive genes as outliers with values < mean
      mutate(local_sensitive = ifelse(local_outlier & LN_IC50 < local_LN_IC50_mean, TRUE, FALSE)) %>%
      mutate(sensitive = local_sensitive) %>%
      separate(target_ID, into = c("dataset", "drug_name", "gene_ID")) %>%
      dplyr::select(dataset, cell_ID, drug_name, gene_ID, LN_IC50, local_LN_IC50_mean, local_outlier_rank, local_outlier, local_sensitive, sensitive)
    
    
    if(global_alpha != FALSE){
      
      lin_drug_sensitivity <- lin_drug_sensitivity %>%
        left_join(GDSC_global_outliers, by = c("cell_ID","gene_ID")) %>%
        mutate(sensitive = ifelse(global_sensitive, TRUE, sensitive))

    }
    
    rm(lin_GDSC_matrix)
  
  }
  
  
  # Dealing with cases of n < 4 for drug sensitivity or gene effect data
  if(length(which(lineage_cells$cell_ID %in% rownames(gene_effect))) < 4){
    #lin_gold_standards <- lin_drug_sensitivity %>% 
    #  dplyr::select(cell_ID, gene_ID, local_sensitive, global_sensitive, sensitive) %>%
    #  mutate(is_gold_standard = sensitive) %>%
    #  filter(is_gold_standard) %>%
    #  unique() %>%
    #  mutate(lineage = lin, dependent = FALSE, local_dependent=FALSE) %>%
    #  dplyr::select(cell_ID,gene_ID,local_dependent, global_dependent,dependent,local_sensitive,global_sensitive,sensitive,is_gold_standard,lineage)
    lin_gene_effect <- gene_effect_global_outliers %>%
      mutate(local_gene_effect_mean = NA, local_outlier_rank = NA, local_outlier = FALSE, local_dependent = FALSE, dependent=global_dependent) %>%
      dplyr::select(cell_ID, gene_ID, gene_effect, local_gene_effect_mean, local_outlier_rank, local_outlier, local_dependent, global_dependent,dependent)
  }else if(length(which(lineage_cells$cell_ID %in% rownames(GDSC))) < 4){
    #lin_gold_standards <- lin_gene_effect %>% 
    #  dplyr::select(cell_ID, gene_ID, local_dependent, global_dependent,dependent) %>%
    #  mutate(is_gold_standard = dependent) %>%
    #  filter(is_gold_standard) %>%
    #  unique() %>%
    #  mutate(lineage = lin, sensitive = FALSE, local_sensitive = FALSE) %>%
    #  dplyr::select(cell_ID,gene_ID,local_dependent, global_dependent,dependent,local_sensitive,global_sensitive,sensitive,is_gold_standard,lineage)
    lin_drug_sensitivity <- GDSC_global_outliers %>%
      mutate(local_LN_IC50_mean = NA, local_outlier_rank = NA, local_outlier = FALSE, local_sensitive = FALSE, sensitive=global_sensitive) %>%
      dplyr::select(dataset, cell_ID, drug_name, gene_ID, LN_IC50, local_LN_IC50_mean, local_outlier_rank, local_outlier, local_sensitive, global_sensitive,sensitive)
  }
    
  #############################
  # Combined Gold Standards
  #############################
  
  lin_gold_standards <- full_join(
    lin_gene_effect %>% dplyr::select(cell_ID, gene_ID,local_dependent, global_dependent, dependent),
    # Because there are many combinations of drugs and drug-targets, just take the max value for sensitivity to avoid duplicates
    lin_drug_sensitivity %>% 
      group_by(cell_ID, gene_ID) %>% 
      summarise(local_sensitive = max(local_sensitive, na.rm = T),global_sensitive = max(global_sensitive, na.rm = T), sensitive = max(sensitive, na.rm = T)) %>%
      ungroup(),
    by = c("cell_ID", "gene_ID")
    ) %>%
    mutate(is_gold_standard = ifelse(dependent, TRUE, ifelse(sensitive, TRUE, FALSE))) %>%
    filter(is_gold_standard) %>%
    unique() %>%
    mutate(lineage = lin) %>%
    dplyr::select(lineage,cell_ID,gene_ID,local_dependent, global_dependent,dependent,local_sensitive,global_sensitive,sensitive,is_gold_standard)
  
  
  
  if(lin == sort(unique(CCLE_sample_info$lineage))[1]){
    gold_standards <- lin_gold_standards
    gene_effect_local_z_scores <- indiv_gene_effect_local_z_scores
  }else{
    gold_standards <- rbind(gold_standards, lin_gold_standards)
    gene_effect_local_z_scores <- rbind(gene_effect_local_z_scores,indiv_gene_effect_local_z_scores)
  }
  
  
}



gold_standards <- gold_standards %>%
  mutate(dependent = ifelse(is.na(dependent), FALSE, dependent),
         sensitive = ifelse(is.na(sensitive), FALSE, sensitive),
         local_dependent = ifelse(is.na(local_dependent), FALSE, local_dependent),
         local_sensitive = ifelse(is.na(local_sensitive), FALSE, local_sensitive),
         global_dependent = ifelse(is.na(global_dependent), FALSE, global_dependent),
         global_sensitive = ifelse(is.na(global_sensitive), FALSE, global_sensitive)
         ) %>%
  unique()
  
write_csv(gold_standards, "validation_data/all_gold_standards.csv")


gene_effect_z_scores <- full_join(gene_effect_global_z_scores, gene_effect_local_z_scores, by = c("cell_ID","gene_ID")) %>%
  dplyr::select(lineage,cell_ID,gene_ID,global_z_score,local_z_score)

write_csv(gene_effect_z_scores, "validation_data/gene_effect_z_scores.csv")




#############################
# Summary Info
#############################

# This summary shows what percentage of the total genes with available information (~17500 genes) the cells have been identified as being sensitive to, on average
# per lineage

summary <- gold_standards %>%
  group_by(lineage,cell_ID) %>%
  summarise(total = n(), drug_sensitive = sum(sensitive, na.rm = T), gene_dependent = sum(dependent, na.rm = T)) %>%
  group_by(lineage) %>%
  summarise(#mean_total_percentile_genes = mean(total)/length(total_genes)*100, 
            #max_total_percentile_genes = max(total)/length(total_genes)*100, 
            #min_total_percentile_genes = min(total)/length(total_genes)*100,
            #mean_gene_dependent_percentile_genes = mean(gene_dependent)/length(total_genes)*100, 
            #mean_drug_sensitive_percentile_genes = mean(drug_sensitive)/length(total_genes)*100,
            mean_total = mean(total),
            mean_dependent = mean(gene_dependent),
            mean_sensitive = mean(drug_sensitive)
  )

write_csv(summary,"validation_data/summary_all_gold_standards.csv")




#test <- read_csv("validation_data/all_gold_standards.csv") %>%
#  filter(gene_ID == "CCNF") %>%
#  left_join(gene_effect_global_outliers, by = c("gene_ID", "cell_ID"))



}












if(method=="picklesv3"){
  



###################
# PICKLES v3
###################
# https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9825567/
# https://pickles.hart-lab.org/

# Recommended thresholds
# BF > 10
# Z < -4
# Chronos < -0.75

  
# Set Thresholds for all gold-standards
z_th <- -4
BF_th <- 10
chronos_th <- -0.75

picklesv3 <- fread("data/PICKLESv3/master_table_Avana_Score_TKOv3_BF_Zscore_Expr_LOF_GOF_18959genes_1162cells_lineage_disease.txt")

ggplot(picklesv3 %>% pivot_longer(cols = c(Avana_Z,Avana_BF,Avana_Chronos,Score_Z,Score_BF,TKOv3_Z,TKOv3_BF), names_to = "measure", values_to = "value"), aes(x=value)) +
  geom_density() +
  facet_wrap(~measure, scales = "free")

ggsave("plots/QC/picklesv3.png", width = 50, height = 50, units = "cm", dpi = 300)


picklesv3_all_gold_standards <- picklesv3 %>%
  dplyr::select(-c(V1,LOF,GOF,primary_disease,lineage,Expr,Cell_Line)) %>%
  dplyr::rename(cell_ID=stripped_cell_line_name, gene_ID=Gene) %>%
  relocate(cell_ID) %>%
  unique() %>%
  filter(cell_ID %in% CCLE_sample_info$cell_ID, gene_ID %in% genes) %>%
  # Add thresholds
  mutate(
    Avana_Z_essential = Avana_Z < z_th,
    Avana_BF_essential = Avana_BF > BF_th,
    Avana_Chronos_essential = Avana_Chronos < chronos_th,
    Score_Z_essential = Score_Z < z_th,
    Score_BF_essential = Score_BF > BF_th,
    TKOv3_Z_essential = TKOv3_Z < z_th,
    TKOv3_BF_essential = TKOv3_BF > BF_th,
  ) %>%
  # Check that all AVAILABLE annotations agree on essentiality
  rowwise() %>%
  mutate(
    consensus_essential = sum(Avana_Z_essential,Avana_BF_essential,Avana_Chronos_essential,Score_Z_essential,Score_BF_essential,TKOv3_Z_essential,TKOv3_BF_essential, na.rm = T) == 
      sum(!is.na(c(Avana_Z_essential,Avana_BF_essential,Avana_Chronos_essential,Score_Z_essential,Score_BF_essential,TKOv3_Z_essential,TKOv3_BF_essential)))
  ) %>%
  ungroup() %>%
  filter(consensus_essential) %>%
  dplyr::select(cell_ID, gene_ID) %>%
  unique() %>%
  left_join(CCLE_sample_info %>% dplyr::select(cell_ID,lineage), by = "cell_ID") %>%
  dplyr::select(lineage, cell_ID, gene_ID)

write_csv(picklesv3_all_gold_standards, "validation_data/all_gold_standards.csv")



# Set thresholds for rare essential genes
z_th <- -3
BF_th <- 8
chronos_th <- -0.5
global_essentiality_freq_thresh <- 0.5
lineage_essentiality_freq_thresh <- 0.25

# Essentiality frequency

picklesv3_rare_gold_standards <- picklesv3 %>%
  dplyr::select(-c(V1,LOF,GOF,primary_disease,lineage,Expr,Cell_Line)) %>%
  dplyr::rename(cell_ID=stripped_cell_line_name, gene_ID=Gene) %>%
  relocate(cell_ID) %>%
  unique() %>%
  filter(cell_ID %in% CCLE_sample_info$cell_ID, gene_ID %in% genes) %>%
  # Add thresholds
  mutate(
    Avana_Z_essential = Avana_Z < z_th,
    Avana_BF_essential = Avana_BF > BF_th,
    Avana_Chronos_essential = Avana_Chronos < chronos_th,
    Score_Z_essential = Score_Z < z_th,
    Score_BF_essential = Score_BF > BF_th,
    TKOv3_Z_essential = TKOv3_Z < z_th,
    TKOv3_BF_essential = TKOv3_BF > BF_th,
  ) %>%
  # Check that all AVAILABLE annotations agree on essentiality
  rowwise() %>%
  mutate(
    consensus_essential = sum(Avana_Z_essential,Avana_BF_essential,Avana_Chronos_essential,Score_Z_essential,Score_BF_essential,TKOv3_Z_essential,TKOv3_BF_essential, na.rm = T) == 
      sum(!is.na(c(Avana_Z_essential,Avana_BF_essential,Avana_Chronos_essential,Score_Z_essential,Score_BF_essential,TKOv3_Z_essential,TKOv3_BF_essential)))
  ) %>%
  ungroup() %>%
  filter(consensus_essential) %>%
  dplyr::select(cell_ID, gene_ID) %>%
  unique() %>%
  left_join(CCLE_sample_info %>% dplyr::select(cell_ID,lineage), by = "cell_ID") %>%
  # Get the total # of cells
  mutate(total_n = length(unique(cell_ID))) %>%
  # Get the global frequency of essentiality
  group_by(gene_ID) %>% mutate(global_essentiality_count = length(unique(cell_ID))) %>% ungroup() %>%
  # Get the total # of cells per lineage
  group_by(lineage) %>% mutate(lineage_n = length(unique(cell_ID))) %>% ungroup() %>%
  # Get the lineage frequency of essentiality
  group_by(gene_ID,lineage) %>% mutate(lineage_essentiality_count = length(unique(cell_ID))) %>% ungroup() %>%
  mutate(global_essentiality_frequency = global_essentiality_count / total_n,
         lineage_essentiality_frequency = lineage_essentiality_count / lineage_n) %>%
  # Filter frequency thresholds
  filter(global_essentiality_frequency < global_essentiality_freq_thresh, lineage_essentiality_frequency < lineage_essentiality_freq_thresh) %>%
  dplyr::select(lineage, cell_ID, gene_ID, global_essentiality_frequency, lineage_essentiality_frequency)

write_csv(picklesv3_rare_gold_standards, "validation_data/rare_gold_standards.csv")
#original_gold_standards <- fread("validation_data/all_gold_standards.csv")

#compare <- original_gold_standards %>%
#  dplyr::select(cell_ID,gene_ID) %>%
#  left_join(picklesv3_gold_standards %>% dplyr::select(cell_ID,gene_ID) %>% mutate(match=TRUE)) %>%
#  mutate(match = ifelse(is.na(match), FALSE, match)) %>%
#  group_by(match) %>%
#  summarise(count=n())


}