# choose_cells.r
# This script prepares all input data for running through the driver prioritisation pipeline

# Load Required Packages
suppressPackageStartupMessages (library(optparse, quietly = T))
suppressPackageStartupMessages (library(tidyverse, quietly = T))
suppressPackageStartupMessages (library(DESeq2, quietly = T))
suppressPackageStartupMessages (library(ggrepel, quietly = T))
suppressPackageStartupMessages (library(data.table, quietly = T))

# Handling input arguments
option_list = list(
  make_option(c("-w", "--workdir"), type="character", default=NULL, 
              help="working directory", metavar ="character"),
  make_option(c("-n", "--network"), type="character", default="STRINGv11", 
              help="network to use", metavar ="character")
); 

opt_parser = OptionParser(option_list=option_list);
opt = parse_args(opt_parser);
if (is.null(opt$workdir)){
  print_help(opt_parser)
  stop("Please provide a working directory", call.=FALSE)
}

# Set the working directory
setwd(opt$workdir)
#setwd("/Users/jc428796/OneDrive - James Cook University(1)/PhD/Bioinformatics/Driver_Prioritisation_Comparison_Pipeline")
#setwd("E:/JCU OneDrive/OneDrive - James Cook University/PhD/Bioinformatics/Driver_Prioritisation_Comparison_Pipeline")

network_choice <- opt$network


#############################
# Get cell list
###############################
# This cell list is generated by the choose_cells.r script

cell_list <- read_csv("data/cell_list.csv")

#############################
# Get cell line information
###############################

CCLE_sample_info <- read_csv("data/CCLE/Model.csv") %>%
  dplyr::rename(DepMap_ID=ModelID, lineage=OncotreeLineage, cell_ID=StrippedCellLineName) %>%
  filter(cell_ID %in% cell_list$cell_ID)

CCLE_model_profile <- read_csv("data/CCLE/OmicsDefaultModelProfiles.csv") %>%
  dplyr::rename(DepMap_ID=ModelID)

# Only retain cell lines with all needed info
CCLE_sample_info <- CCLE_sample_info %>%
  filter(DepMap_ID %in% intersect(CCLE_sample_info$DepMap_ID, CCLE_model_profile$DepMap_ID))
CCLE_model_profile <- CCLE_model_profile %>%
  filter(DepMap_ID %in% intersect(CCLE_sample_info$DepMap_ID, CCLE_model_profile$DepMap_ID))

# Creating a df of DepMap ID / Cell Name mapping
CCLE_IDs <- CCLE_sample_info %>%
  dplyr::select(DepMap_ID, cell_ID)

CCLE_model_profile <- CCLE_model_profile %>%
  inner_join(CCLE_IDs, by = c("DepMap_ID"))

#########################
# Filtering Expression data based on TPM to only include expressed genes
#########################

CCLE_tpm <- data.table::fread("data/CCLE/OmicsExpressionProteinCodingGenesTPMLogp1.csv") %>%
  # Replacing DepMap IDs with cell names
  dplyr::rename(DepMap_ID = 1) %>%
  dplyr::right_join(CCLE_IDs, by = "DepMap_ID") %>%
  na.omit() %>%
  dplyr::select(-DepMap_ID) %>% dplyr::relocate(cell_ID) %>%
  data.table::transpose(make.names = "cell_ID", keep.names = "gene_ID") %>%
  # Fixing gene_IDs by removing everything after a space
  mutate(gene_ID = gsub(" .*","",gene_ID)) %>%
  arrange(gene_ID) %>%
  column_to_rownames("gene_ID") %>%
  # Trimming genes with low reads
  
  # Filtering low expression based on TPM as this accounts for gene length
  # Cells contain approx 200 000 mRNA at any time, therefore tpm of 5 ~ 1 transcript per cell
  # Therefore, tpm of 5 is beging used a cutoff to select genes that are being expressed
  # Data is in log2(tpm+1), therefore filtering for expression abofe log2(5+1)
  dplyr::filter(rowMeans(select(., where(is.numeric))) > log2(5+1)) %>%
  as.data.frame()

CCLE_counts <- data.table::fread("data/CCLE/OmicsExpressionGenesExpectedCountProfile.csv") %>%
  dplyr::mutate(across(-1, round)) %>%
  # Replacing DepMap IDs with cell names
  dplyr::rename(ProfileID = 1) %>%
  dplyr::right_join(CCLE_model_profile, by = "ProfileID") %>%
  na.omit() %>%
  dplyr::select(-c(ProfileID,ProfileType,DepMap_ID)) %>% dplyr::relocate(cell_ID) %>%
  # Fixing gene_IDs by removing everything after a space
  data.table::transpose(make.names = "cell_ID", keep.names = "gene_ID") %>%
  as.data.frame() %>%
  mutate(gene_ID = gsub(" .*","",gene_ID)) %>%
  dplyr::filter(gene_ID %in% rownames(CCLE_tpm))

# Removing duplicate genes
dup <- rle(sort(CCLE_counts$gene_ID))
CCLE_counts <- CCLE_counts %>%
  filter(gene_ID %in% dup$values[dup$lengths==1])

rm(dup)

CCLE_counts <- CCLE_counts %>%
  column_to_rownames("gene_ID")

CCLE_tpm <- CCLE_tpm[rownames(CCLE_counts),]


#########################
# Network
#########################

if(network_choice=="STRINGv11"){
  
}
